lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,208 @@
+      # and the BatchWriter can be inspected to see what the errors were.
+      # A call to #errors will return an array of response objects that contain
+      # the errors. See the documentation for Response at
+      # http://docs.aws.amazon.com/sdkforruby/api/Seahorse/Client/Response.html
+      #
+      # @param [Hash] opts options to pass on to the client call to
+      #   +#batch_item_write+. See the documentation above in the AWS SDK for
+      #   Ruby V2.
+      # @opts retry_count the number of retries to attempt on failed write
+      # @raise RuntimeError if unsuccessful in submitting all items
+      def save!(opts = {})
+        raise "can't submit if writer in #{@state} state." unless pending?
+        validate_items
+        retry_items = send_chunks(opts)
+        responses = retry_failed_requests(retry_items, opts)
+        analyze_result(responses)
+      rescue RuntimeError => e
+        capture_validation_error(e)
+        raise e
+      end
+
+      # Perform the save! function but capture the error that is raised on
+      # unsuccessful save.
+      def save(opts = {})
+        save!(opts)
+        true
+      rescue RuntimeError
+        false
+      end
+
+      def error?
+        @state == ERROR_SEND_STATE
+      end
+
+      # Get a hash of the responses that didn't work along with their
+      # unprocessed items. Mimics ActiveRecord design.
+      def errors
+        @errors.map do |err|
+          {
+            error: err.error,
+            data: err.data,
+            unprocesssed_items: err.unprocessed_items
+          }
+        end
+      end
+
+      def pending?
+        @state == PENDING_STATE
+      end
+
+      def unprocessed_items
+        @errors.map(&:unprocessed_items)
+      end
+
+      def valid?
+        @state == SUCCESSFUL_SEND_STATE
+      end
+
+      private
+      def alert_errors(responses)
+        @errors += responses.reject(&:successful?)
+        @state = ERROR_SEND_STATE
+        raise Errors::SubmissionError
+      end
+
+      def analyze_result(responses)
+        alert_errors(responses) if failures?(responses)
+        @state = SUCCESSFUL_SEND_STATE
+      end
+
+      def batch_write(item_requests_array, opts)
+        @client.batch_write_item(
+          formatted_request(item_requests_array, opts)
+        )
+      end
+
+      def batch_write_items(item_requests_array, opts)
+        response = batch_write(item_requests_array, opts)
+        if response.successful?
+          response.unprocessed_items[@model.table_name] || []
+        else
+          item_requests_array
+        end
+      end
+
+      def capture_validation_error(e)
+        @state = ERROR_SEND_STATE
+        error = Struct.new(:error, :data, :unprocessed_items)
+        @errors << error.new(e.class, e, @items)
+      end
+
+      def failures?(responses)
+        responses.any? do |response|
+          !response.successful? || response.unprocessed_items.empty?
```

Comment:
A write fails if it returns a ~400 error code or it fails to write all of the items (unprocessed_items). It would be hard to really convey success and failure with a single status code, so we have to check if any unprocessed_items exist, and if they do we know some of the writes have failed.